### PR TITLE
Add Microsoft xoauth2 support

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/mail/issues/6591

## Todo
- [x] Requires https://github.com/nextcloud/mail/pull/7721
- [x] Oauth workflow
- [x] User UI
- [x] Admin UI

### CODE REVIEW

**Comment by ChristophWurst** [comment]
> User UI

I'm facing a little challenge here.

Generally speaking, IMAP accounts with password auth require us to ask the user for a password. XOAUTH2 accounts do not need a password but use a popup screen for the vendor.

With the Gmail XOAUTH2 integration I chose to hide the password input as soon as someone enters an email address that ends with `@gmail.com`. This works. But it only works for Gmail accounts with the default Gmail domain. Any custom domain hosted by Google falls short. This was acceptable for a first version.

Now, with Microsoft, it seems even more likely that someone uses a domain that doesn't easily allow us to detect it's hosted by Microsoft. Some accounts seem to end with `@onmicrosoft.com` but there are many more.

A more sophisticated approach for detecting Gmail and Microsoft accounts is to use the IMAP and SMTP hosts found by auto config detection. But auto config detection only happens *after* the user clicks on *Connect*. At that point we should already have hidden the password, or not.

So there are at least two approaches we can take here.

1) Always show the password input, but no longer make it required for auto mode. Manual mode stays as is. 
    If we detect that the account uses password auth then we abort the setup right after auto detection
    If we detect that it's a Gmail or Microsoft account we just discard the password.
2) Change the setup into a multi step dialogue
    1. Ask for the name and email address
    2. Perform auto detection in the background
    3. If XOAUTH2, open the vendor popup
    3. If config for a password account could be detected, ask for the password
    3. If no config could be detected, show the manual setup

For simplicity I'll probably go with the first one.

**Comment by ChristophWurst** [comment]
Tada

Bildschirmfoto vom 2023-01-19 11-39-35

**Comment by ChristophWurst** [comment]
Until it is done

**Comment by ChristophWurst** [comment]
Please be patient and refrain from commenting on PRs unless the comment adds value to the code changes or discussion. We are struggling too many notifications already. The comments just slow us down. Thank you.

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  -->
+
+<template>
+	<form @submit.prevent="onSubmit">
+		<label for="mail-gmail-oauth-tenant-id"> {{ t('mail', 'Tenant ID') }} </label>
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:24]:
```suggestion
		<label for="mail-microsoft-oauth-tenant-id"> {{ t('mail', 'Tenant ID') }} </label>
```

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		<input
+			id="mail-gmail-oauth-tenant-id"
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:26]:
```suggestion
			id="mail-microsoft-oauth-tenant-id"
```

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			v-model="tenantIdVal"
+			:disabled="loading"
+			type="text"
+			required>
+		<label for="mail-gmail-oauth-client-id"> {{ t('mail', 'Client ID') }} </label>
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:31]:
```suggestion
		<label for="mail-microsoft-oauth-client-id"> {{ t('mail', 'Client ID') }} </label>
```

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		<input
+			id="mail-gmail-oauth-client-id"
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:33]:
```suggestion
			id="mail-microsoft-oauth-client-id"
```

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+			v-model="clientIdVal"
+			:disabled="loading"
+			type="text"
+			required>
+		<label for="mail-gmail-oauth-client-secret"> {{ t('mail', 'Client secret') }} </label>
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:38]:
```suggestion
		<label for="mail-microsoft-oauth-client-secret"> {{ t('mail', 'Client secret') }} </label>
```

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		<input
+			id="mail-gmail-oauth-client-secret"
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:40]:
```suggestion
			id="mail-microsoft-oauth-client-secret"
```

File: src/components/settings/MicrosoftAdminOauthSettings.vue (hunk 1 above)
Comment (st3iny) [reply src/components/settings/MicrosoftAdminOauthSettings.vue:24]: Styling might be adjusted. Or is it just used for the label? (Applies to the other occurrences as well.)

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+			v-model="clientSecret"
+			:disabled="loading"
+			type="password"
+			required>
+		<button type="submit" :disabled="!tenantId || !clientIdVal || !clientSecret || clientSecret === PASSWORD_PLACEHOLDER || loading" class="primary">
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:45]:
```suggestion
		<button type="submit" :disabled="!tenantIdVal || !clientIdVal || !clientSecret || clientSecret === PASSWORD_PLACEHOLDER || loading" class="primary">
```

Fixes save button never being enabled.

File: src/components/settings/MicrosoftAdminOauthSettings.vue [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+			clientIdVal: this.clientId,
+			clientSecret: this.clientId ? PASSWORD_PLACEHOLDER : '',
+			PASSWORD_PLACEHOLDER,
+		}
+	},
+	methods: {
+		async onSubmit() {
+			this.loading = true
+			try {
+				await configure(this.tenantId, this.clientIdVal, this.clientSecret)
```
Comment (st3iny) [inline src/components/settings/MicrosoftAdminOauthSettings.vue:88]:
```suggestion
				await configure(this.tenantIdVal, this.clientIdVal, this.clientSecret)
```

Fixes tenant id never being sent to the backend.

File: src/components/settings/MicrosoftAdminOauthSettings.vue (hunk 7 above)
Comment (st3iny) [reply src/components/settings/MicrosoftAdminOauthSettings.vue]: Done

File: appinfo/routes.php [hunk 9]
```diff
@@ -369,6 +369,21 @@
...
 			'verb' => 'GET',
 		],
+		[
+			'name' => 'microsoftIntegration#configure',
+			'url' => '/api/integration/microsoft',
+			'verb' => 'POST',
+		],
+		[
+			'name' => 'microsoftIntegration#unlink',
+			'url' => '/api/integration/google',
```
Comment (kesselb) [inline appinfo/routes.php:379]:
```suggestion
			'url' => '/api/integration/microsoft',
```

;)

**Review by kesselb — APPROVED** [review]

**Review by st3iny — APPROVED** [review]
Tested and works with my personal account.

**Comment by flowlee** [comment]
This does not currently work for addresses with the @bildung.gv.at domain. These are the official education addresses used by teachers in Austria and they are hosted by Microsoft and normally accessed via outlook.com.

**Pull request merged**